### ISSUE TITLE
Filtering Recommendations does not include type

### ISSUE THREAD

**Issue by ColtonProvias** [json-api/json-api#389]
http://jsonapi.org/recommendations/#filtering

The filtering recommendations provided assume only ids are used for referencing related resources.  However, this conflicts with the current spec in which we have related resources being related via both id and type.

**Comment by tkellen** [comment]
JSON-API does not define the semantics of `filter`. There is nothing stopping an implementor from doing `filter[TYPE.field]` or similar. I'm closing for now, but I would love a PR adding this to the filtering section of the spec @ColtonProvias!

**Comment by ethanresnick** [comment]
I agree that the spec leaves this behavior undefined (and should continue to!), but I do think we could make the examples on the recommendations page better, since @ColtonProvias's is right that they currently act as if polymorphic collections don't exist (by filtering only on a list of ids). I'm gonna reopen this just as a reminder to self (or someone else) to write up a PR.

**Comment by ColtonProvias** [comment]
Since the standard of types in the spec seems to be underscore or camel-cased, maybe the best way to represent type-id pairs would then be to separate them by a dot.  For example, in SQLAlchemy-JSONAPI, I represent individual resources in errors as `type.id` (for example: `users.3f35f2b0-2c31-4624-97fa-b5cff53128b3`).  That might be a way to go about it.

The real hassle with filtering, as I'm finding, is locating a balance between moderately-simple filters like `filter[published_at.gt]=2015-09-21T17%3A09%3A20.380443%2B00%3A00&filter[author]=users.3f35f2b0-2c31-4624-97fa-b5cff53128b3` and complex filters that include logical ORs and such.  Of course, that's probably a discussion best left up for the implementation of the backend, but should still probably be standardized somehow to make it easier on clients.

**Comment by ethanresnick** [comment]
@ColtonProvias I like the dot quite a lot. Would you be willing to whip up a PR to add that to these filtering examples? If so, I'd gladly merge it in

**Comment by ColtonProvias** [comment]
Sorry for the delay, but right now I'm working on getting it setup within SQLAlchemy-JSONAPI.  I posted a quick write-up of it at https://github.com/ColtonProvias/sqlalchemy-jsonapi/wiki/Filtering if you guys want to take a look.

**Comment by zeorin** [comment]
@ColtonProvias I know this is nearly a year later, but your write up of filtering really helped me out. It's a shame I had to search GitHub issues to find it, though. I think this is be a very reasonable recommendation about filtering strategies.

**Comment by jelhan** [comment]
I agree that the current filter recommendation may be misleading as it does not support polymorphic collections. But to be honest I had the feeling that the filtering recommendation is way over-simplified to serve real use cases. Rather than adding on it, I would recommend to write some fully-fledged profiles for filtering. One or more of them could be referenced as a recommendation later.

**Comment by freddrake** [comment]
@jelhan: The current specifications only state that the ``filter`` query parameter be used to provide a filter; everything else is up to the application.

There is nothing in the specification that prohibits polymorphic collections.

**Comment by jelhan** [comment]
> @jelhan: The current specifications only state that the `filter` query parameter be used to provide a filter; everything else is up to the application.
> 
> There is nothing in the specification that prohibits polymorphic collections.

Fully agree. But this issue is about the [filtering _recommendation_](https://jsonapi.org/recommendations/#filtering) if I got it right.

**Comment by freddrake** [comment]
@jelhan: Good point; you're right.

I've generally ignored the recommendations page because they seem as narrowly focused as the examples in the specification; useful for illustration, but not really helpful on their own.
